src/madt.rs

```diff
@@ -0,0 +1,469 @@
+        size: u32,
+        checksum: u8,
+        oem_id: [u8; 6],
+        oem_table_id: [u8; 8],
+        oem_revision: u32,
+        creator_id: u32,
+        creator_revision: u32,
+    ) -> Madt {
+        Madt {
+            header: SdtHeader::make_testcase(
+                *b"APIC",
+                size,
+                45,
+                checksum, // checksum
+                oem_id,
+                oem_table_id,
+                oem_revision,
+                creator_id,
+                creator_revision,
+            ),
+            local_interrupt_controller_address: 0xDEADBEEF as u32,
+            flags: 0 as u32,
+        }
+    }
+}
+
+pub fn parse_madt(mapping: &PhysicalMapping<Madt>) -> Result<(), AcpiError> {
+    (*mapping).validate()?;
+
+    let mut i: usize = mem::size_of::<Madt>();
+    while i < (*mapping).length() as usize {
+        let struct_type: u8 =
+            unsafe { *((mapping.virtual_start.as_ptr() as *const u8).add(i) as *const u8) } as u8;
+        let length: u8 = unsafe {
+            *((mapping.virtual_start.as_ptr() as *const u8).add(i + 1) as *const u8)
+        } as u8;
+        match struct_type {
+            MADT_PROCESSOR_LOCAL_APIC => unsafe {
+                (*((mapping.virtual_start.as_ptr() as *const u8).add(i)
+                    as *const MadtProcessorLocalAPIC))
+                    .validate()?;
+            },
+            MADT_IO_APIC => unsafe {
+                (*((mapping.virtual_start.as_ptr() as *const u8).add(i) as *const MadtIOAPIC))
+                    .validate()?;
+            },
+            MADT_INTERRUPT_SOURCE_OVERRIDE => unsafe {
+                (*((mapping.virtual_start.as_ptr() as *const u8).add(i)
+                    as *const MadtInterruptSourceOverride))
+                    .validate()?;
+            },
+
+            MADT_LOCAL_APIC_NMI => unsafe {
+                (*((mapping.virtual_start.as_ptr() as *const u8).add(i)
+                    as *const MadtLocalAPICNMI))
+                    .validate()?;
+            },
+
+            _ => {
+                unimplemented!();
+            }
+        }
+        i = i + (length as usize);
+    }
+
+    Ok(())
+}
+
+const MADT_PROCESSOR_LOCAL_APIC: u8 = 0;
+const MADT_PROCESSOR_LOCAL_APIC_LENGTH: u8 = 8;
+
+#[repr(C, packed)]
+struct MadtProcessorLocalAPIC {
+    struct_type: u8, // 0
+    length: u8,      // 8
+    acpi_processor_uid: u8,
+    apic_id: u8,
+    flags: u32,
+}
+
+impl MadtProcessorLocalAPIC {
+    fn validate(&self) -> Result<(), AcpiError> {
+        // TODO change the errors
+        // Also is it also needed to check struct type and length ??
+        if self.struct_type != 0 {
+            return Err(AcpiError::Error);
+        }
+        if self.length != 8 {
+            return Err(AcpiError::Error);
+        }
+        if self.flags.get_bits(1..31) != 0 {
+            return Err(AcpiError::Error);
+        }
+        Ok(())
+    }
+
+    fn length() -> u32 {
+        8 as u32
```

Comment:
What is the purpose of this method? Surely it is just `mem::size_of::<MadtProcessorLocalApic>()`.